R/functions.R

```diff
@@ -457,6 +445,90 @@ m.auc <- function(data.m, group.v) {
   return(AUC)
 }
 
+# Functions for MAST
+
+createLrTestTable <- function(i,de, zlm.lr_pvalue) {
+  lrTest.table <-
+    merge(zlm.lr_pvalue, de, by.x = "primerid", by.y = "row.names")
+  colnames(lrTest.table) <-
+    c(
+      "gene",
+      "test.type",
+      "p_value",
+      paste("log2.mean.", "Cluster_Other", sep = ""),
+      paste("log2.mean.", i, sep = ""),
+      "log2fold_change",
+      "Auc"
+    )
+  return (lrTest.table[rev(order(lrTest.table$Auc)), ])
+}
+saveMastResult <- function(path, cluster_lrTest.table, i) {
+  if (!is.null(path)) {
+    utils::write.csv(cluster_lrTest.table,
+                     file = paste(path, "/", i, "_lr_test.csv", sep = "")
+    )
+    save(cluster_lrTest.table,
+         file = paste(path, "/", i, "_mist.RData", sep = "")
+    )
+  }
+}
+
+doMastZlm <- function(verbose, vbeta.1) {
+  zlm.output <-
+    verbose_wrapper(verbose)(MAST::zlm(
+      ~Population,
+      vbeta.1,
+      method = "bayesglm",
+      ebayes = TRUE
+    ))
+  if (verbose) {
+    methods::show(zlm.output)
+    coefAndCI <- summary(zlm.output, logFC = TRUE)
+    print(coefAndCI)
+  }
+  return (zlm.output)
+}
+
+doLrTest <- function(verbose, zlm.output) {
+  zlm.lr <-
+    verbose_wrapper(verbose)(MAST::lrTest(zlm.output, "Population"))
+  zlm.lr_pvalue <- data.table::melt(zlm.lr[, , "Pr(>Chisq)"])
+  zlm.lr_pvalue <-
+    zlm.lr_pvalue[which(zlm.lr_pvalue$test.type == "hurdle"), ]
+  return (zlm.lr_pvalue)
+}
+
+createDataForMAST <- function(verbose, counts, groups) {
+  data_for_MIST <-
+    verbose_wrapper(verbose)(as.data.frame(cbind(
+      rep(rownames(counts), dim(counts)[2]),
+      data.table::melt(counts),
+      rep(groups, each = dim(counts)[1]),
+      rep(1, dim(counts)[1] * dim(counts)[2])
+    )))
+  colnames(data_for_MIST) <- c(
+    "gene",
+    "Subject.ID",
+    "Et",
+    "Population",
+    "Number.of.Cells"
+  )
+  vbeta.fa <-
+    verbose_wrapper(verbose)(
+      MAST::FromFlatDF(
+        data_for_MIST,
+        idvars = c("Subject.ID"),
+        primerid = "gene",
+        measurement = "Et",
+        ncells = "Number.of.Cells",
+        geneid = "gene",
+        cellvars = c("Number.of.Cells", "Population"),
+        phenovars = c("Population"),
+        id = "vbeta all"
+      )
+    )
+  return(subset(vbeta.fa, Number.of.Cells == 1))
+}
```

Comment:
in this function there would be the potential to do some additional speed up by parallelization. But then the verbose wrapper would not print anything

Reply:
i would say speed > verbosity in this case.

Reply:
difficult question, since it also would suppress error messaging which makes the library harder to use

Reply:
Thats true. I think Lorenzo would also like to have a faster DWLS, since building four signatures took him over 5 days :D

But we should maybe discuss this in the group.

Reply:
I found a better way to speed up this function by using MAST::fromMatrix instead of MAST::fromFlatDF.

I also tried to use parallelisation here, but then the memory consumption explodes, because very large files are duplicated on every core. this killed computer 😄